---
title: "20241020_DESeq2_volcano"
output: html_document
date: "2024-10-20"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library( "DESeq2" )
library(ggplot2)
library(pheatmap)
library(dplyr)
library(curatedMetagenomicData)
library(stringr)
theme_set(theme_bw())
```

## DESeq2 identify differential abundance species

Here we use DESeq2 to detect differential abundance species between __healthy__ and __non-healthy__ samples.

The examples in this RMD use the following data set: `20241006_diarrhea_v5_counts.rds`

Here we set `sfType = "poscounts"` for the function `estimateSizeFactors()`. For this function, available method for estimation: either "ratio", "poscounts", or "iterate". "ratio" uses the standard median ratio method introduced in DESeq. The size factor is the median ratio of the sample over a "pseudosample": for each gene, the geometric mean of all samples. "poscounts" and "iterate" offer alternative estimators, which can be used even when all genes contain a sample with a zero (a problem for the default method, as the geometric mean becomes zero, and the ratio undefined). The "poscounts" estimator deals with a gene with some zeros, by calculating a modified geometric mean by taking the n-th root of the product of the non-zero counts. This evolved out of use cases with Paul McMurdie's phyloseq package for __metagenomic samples__. More details can be found [here](https://support.bioconductor.org/p/9160016/#9160017).

```{r deseq, echo=TRUE, warning=FALSE, message=FALSE, eval = FALSE}
  dia_v5_ra <- readRDS("20241006_diarrhea_v5_counts.rds")
  #dia_v5_ra <- readRDS("20240928_RG_diarrhea_v5.rds")
  
  colData(dia_v5_ra)$Disease <- ifelse(colData(dia_v5_ra)$disease == "healthy", "Healthy", "Non-healthy") #CAPITAL D is what we want
  ## construct DESeq DataSet Object
  dds <- DESeqDataSet(se = dia_v5_ra, design=~Disease)
  ds = DESeq(dds, fitType="local", sfType = "poscounts")
  saveRDS(ds, "20241020_DESeq_disease.rds")
```

## Volcano plot of differentially abundant species.

Then we plot the results generated by DESeq2. 

```{r vol1, echo=FALSE, warning=FALSE, message=FALSE, fig.width=5, fig.height=4}
  ds <- readRDS("20241020_DESeq_disease.rds")
  #ds <- readRDS("20241006_DESeq_disease.rds")
  res_ds <- results(ds, alpha = 0.05)
  res_ds[is.na(res_ds$padj),] <- 1

  p <- ggplot(res_ds, aes(x = log2FoldChange, y = -log10(padj))) +
     geom_point(aes(colour = padj < 0.01), size=1) +
     labs(x = "log2(fold change)", y = "-log10(p-value)", colour = "FDR < 0.01",
                   title = "Non-healthy vs healthy") +
     theme_bw() +
     scale_color_manual(values = c("gray", "skyblue2"))
  
  # library (EnhancedVolcano)
  # EnhancedVolcano(res_ds,
  #   lab = rownames(res_ds),
  #   x = 'log2FoldChange',
  #   y = "padj",
  #   #y = 'pvalue',
  #   pCutoff = 0.01, #p-value cutoff <0.01
  #   pointSize = 3.0,
  #   labSize = 6.0)
  ggsave("DESeq2_Speciec.png", plot=p, height=6, width=8, dpi="print")
```

To have a basic idea about how many species are significantly more abundant in non-healthy and vice verse, we count the number of species as follow:

Species that are significantly (adjust P value < 0.01) higher in non-healthy population:
`sum(res_ds$log2FoldChange > 0 & res_ds$padj < 0.01)` = `r sum(res_ds$log2FoldChange > 0 & res_ds$padj < 0.01)`. This is the number of blue points on the right side of the volcano plot.

Species that are significantly (adjust P value < 0.01) higher in healthy population:
`sum(res_ds$log2FoldChange < 0 & res_ds$padj < 0.01)` = `r sum(res_ds$log2FoldChange < 0 & res_ds$padj < 0.01)`.

We also notice that in two sides of the plot, they are some species that are extremely differently abundant (i.e. didn't detect in the other condition). We quickly count the number of those species too.

Species that are only present in non-healthy samples (far right side of volcano):
`sum(res_ds$log2FoldChange > 20 & res_ds$padj < 0.01)` = `r sum(res_ds$log2FoldChange > 20 & res_ds$padj < 0.01)`. 

Species that are only present in healthy samples (far left side of volcano):
`sum(res_ds$log2FoldChange < -20 & res_ds$padj < 0.01)` = `r sum(res_ds$log2FoldChange < -20 & res_ds$padj < 0.01)`. 

We then examine the counts of species across the groups.

```{r counts_cmp, echo=TRUE, warning=FALSE, message=FALSE, eval=TRUE}
  e_coli <- "k__Bacteria|p__Proteobacteria|c__Gammaproteobacteria|o__Enterobacterales|f__Enterobacteriaceae|g__Escherichia|s__Escherichia_coli"
  plotCounts(ds, gene=e_coli, intgroup="Disease")
```

#### Assignment 1 Compare the aggregated/accumulated relative abundance (aRA) of differentially abundant species between groups.

```{r Prep_data, warning=FALSE, echo=FALSE, message=FALSE}
  ra <- assay(dia_v5_ra)
  ra <- as.matrix(apply(ra, 2, function(x) x/sum(x)))
  metadata <- colData(dia_v5_ra)
  metadata$Sample_ID <- rownames(metadata)
  taxa <- str_split_fixed(rownames(ra), "\\|", 7)
  colnames(taxa) <- c("Kingdom", "Phylum", "Class", "Order", "Family",
                            "Genus", "Species")
  rownames(taxa) <- rownames(ra)
```


```{r healthy_aRA, warning=FALSE, message=FALSE, echo=FALSE}
healthy_sig <- res_ds[res_ds$log2FoldChange < -1 & res_ds$padj < 0.01, ] 

#identify differentially abundant healthy species from log2foldchange
healthy_result <- as.data.frame(healthy_sig) %>%
  arrange(log2FoldChange)
#log2FoldChange = -30.00
filtered_healthy_result <- healthy_result %>%
  filter(log2FoldChange == -30)
dim(filtered_healthy_result)
# 20 6
# breakdown the taxa names
taxa <- str_split_fixed(rownames(filtered_healthy_result), "\\|", 7)
colnames(taxa) <- c("Kingdom", "Phylum", "Class", "Order", "Family",
                            "Genus", "Species")
rownames(taxa) <- rownames(filtered_healthy_result)
taxa <- as.data.frame(taxa)
filtered_healthy_result <- as.data.frame(filtered_healthy_result)
filtered_healthy_result <- merge(filtered_healthy_result, taxa, by = 0, all = TRUE)
filtered_healthy_result |> arrange(baseMean)
#%>% head(filtered_healthy_result)[1:5, ]

healthy_list <- healthy_sig@rownames
healthy_ra <- ra[rownames(ra) %in% healthy_list, ]
dim(healthy_ra)
# 85 3112
healthy_aRA_df <- data.frame(Sample_ID = colnames(healthy_ra),
                                 aRA = colSums(healthy_ra))
metadata$Disease <- ifelse(metadata$disease == "healthy", "Healthy", "Non-healthy")
healthy_aRA_df <- merge(metadata, healthy_aRA_df)
  
ggplot(healthy_aRA_df, aes(Disease, aRA)) +
    geom_boxplot() 
  ## Non-healthy vs healthy --> when log2FoldChange > 1, means higher RA in Non-healthy
```

>What is the aRA distribution of species that are more abundant in healthy samples?
Bacteroides vulgatus has aRA of 354.094, followed by Bacteroides fragilis with 58.819 aRA, and Escherichia coli with aRA of 56.474 aRA. 

```{r unhealthy_aRA, warning=FALSE, message=FALSE, echo=FALSE}
unhealthy_sig <- res_ds[res_ds$log2FoldChange > 1 & res_ds$padj < 0.01, ]

unhealthy_result <- as.data.frame(unhealthy_sig) %>%
  arrange(desc(log2FoldChange))
head(unhealthy_result)
#log2FoldChange = 30.00
filtered_unhealthy_result <- unhealthy_result %>%
  filter(log2FoldChange == 30)
dim(filtered_unhealthy_result)
# 25 6
# breakdown the taxa names
taxa <- str_split_fixed(rownames(filtered_unhealthy_result), "\\|", 7)
colnames(taxa) <- c("Kingdom", "Phylum", "Class", "Order", "Family",
                            "Genus", "Species")
rownames(taxa) <- rownames(filtered_unhealthy_result)
taxa <- as.data.frame(taxa)
filtered_unhealthy_result <- as.data.frame(filtered_unhealthy_result)
filtered_unhealthy_result <- merge(filtered_unhealthy_result, taxa, by = 0, all = TRUE)
filtered_unhealthy_result |> arrange(baseMean)


unhealthy_list <- unhealthy_sig@rownames
unhealthy_ra <- ra[rownames(ra) %in% unhealthy_list, ]
dim(unhealthy_ra)

# 121 3112
unhealthy_aRA_df <- data.frame(Sample_ID = colnames(unhealthy_ra),
                                 aRA = colSums(unhealthy_ra))
unhealthy_aRA_df <- merge(metadata, unhealthy_aRA_df)
  
ggplot(unhealthy_aRA_df, aes(Disease, aRA)) +
    geom_boxplot() 
## Non-healthy vs healthy --> when log2FoldChange < 1, means lower RA in Non-healthy

```


>What is the aRA distribution of species that are more abundant in non-healthy samples? 
In non-healthy samples, Ruminococcus bromii has the highest aRA of 40.079, followed by Akkermansia muciniphia with aRA of 31.78, and Eubacterium siraeum with 21.665 aRA. 

>What is the main phyla in those differentially abundant species?
For the healthy samples, the top differentially abundant phyla are Bacteroidetes, Proteobacteria, and actinobacteria. For unhealthy samples, the top differentially abundant phyla are Firmicutes, Verrucomicrobia, and Bacteroidetes.  

```{r heatmap, eval=FALSE, message=TRUE, fig.width=40, fig.height=50}
  diff_list <- c(healthy_list, unhealthy_list)
  diff_ra <- ra[rownames(ra) %in% diff_list, ]
  
  this_taxa <- taxa[rownames(taxa) %in% rownames(diff_ra), ]
  taxa_map <- data.frame(Full_name = rownames(this_taxa),
                         Phylum = this_taxa[, 2],
                         Species = this_taxa[, 7])
  taxa_map <- taxa_map[match(taxa_map$Full_name, rownames(diff_ra)), ]
  #rownames(sorted_healthy_ra) <- colnames(sorted_healthy_ra) <- taxa_map$Species
  
  taxa_map2 <- data.frame(Phylum = factor(x = taxa_map$Phylum,
                                            levels = unique(taxa_map$Phylum)))
  rownames(taxa_map2) <- taxa_map$Species
  
  ## Define colors
  palette_OkabeIto <- c("#E69F00", "#56B4E9", "#009E73", "#F0E442",
                      "#0072B2", "#D55E00", "#CC79A7", "#999999")
  mat_colors <- list(Phylum = palette_OkabeIto[1:length(unique(taxa_map$Phylum))])
  names(mat_colors$Phylum) <- unique(taxa_map2$Phylum)
  
  pheatmap(diff_ra, annotation_col = taxa_map2, annotation_row = taxa_map2,
           annotation_colors = mat_colors, na_col="white",
          border_color ="NA",  treeheight_row = 0, treeheight_col = 0)
```
##WHAT THE DUCK

>Try to plot heatmap of those different abundance species in healthy and non-healthy samples.

## DESeq2 identify differential abundance species with more than two levels 

```{r disease_tidy, echo=TRUE, warning=FALSE, message=FALSE, eval = TRUE}
  dia_v5_ra <- readRDS("20241006_diarrhea_v5_counts.rds")

  ## merge "IBD;perianal_fistula" to "IBD"
  ## merge "CDI;pneumonia", "CDI;cellulitis", "CDI;osteoarthritis", "CDI;ureteralstone"  to "CDI"
  cdi_list <- c("CDI;pneumonia", "CDI;cellulitis", "CDI;osteoarthritis", "CDI;ureteralstone")
  meta <- as.data.frame(colData(dia_v5_ra))
  
  meta <- mutate(meta, disease_tidy = case_when(disease == "IBD;perianal_fistula" ~ "IBD",
                                                disease %in% cdi_list ~ "CDI",
                                                TRUE ~ disease))
  ra <- assay(dia_v5_ra)
```


After getting differentially abundant species between two levels: __healthy__ and __non-healthy__, we extend our model to multiple levels. Since we have too many disease conditions (including `r unique(colData(dia_v5_ra)$disease)`), we will clean the data by tidying up the disease names and limit the samples to the following conditions: `filter_lst <- c("acute_diarrhoea", "CDI", "healthy", "IBD")`. 

```{r deseq_multi, echo=TRUE, warning=FALSE, message=FALSE, eval = FALSE}
  filter_lst <- c("healthy", "IBD", "acute_diarrhoea", "CDI")
  
  meta <- meta[meta$disease_tidy %in% filter_lst, ]
  ra <- ra[, colnames(ra) %in% rownames(meta)]
  ra <- ra[rowSums(ra) > 5, ]
  # 1015 3168
  meta$disease_tidy <- factor(meta$disease_tidy, levels = c("healthy", "acute_diarrhoea", "CDI", "IBD"))

  ## construct DESeq DataSet from matrix
  dds_multi <- DESeqDataSetFromMatrix(countData=ra, 
                              colData=meta, 
                              design=~disease_tidy)

  ds_multi = DESeq(dds_multi, fitType="local", sfType = "poscounts", parallel = T)

  saveRDS(ds_multi, "20241020_DESeq_disease_multi_levels.rds")
```

We then check the results of this new model with multiple levels. Different comparisons were calculated in one command and we can check all comparison using `resultsNames(ds_multi)`. Below we plot the volcano of IBD healthy control samples.

```{r deseq_multi_res, echo=FALSE, warning=FALSE, message=FALSE, eval = FALSE}
  ds_multi <- readRDS("20241020_DESeq_disease_multi_levels.rds")
  resultsNames(ds_multi)
  res_ds_multi <- results(ds_multi, alpha = 0.05)
```

#### Assignment 2 Plot other comparisons of the DESeq.

```{r vol_ad_h, echo=FALSE, warning=FALSE, message=FALSE, fig.width=5, fig.height=4}
  res1_ah <- results(ds_multi, name=("disease_tidy_acute_diarrhoea_vs_healthy"), alpha = 0.05)
  res1_ah[is.na(res1_ah$padj),] <- 1

  ggplot(res1_ah, aes(x = log2FoldChange, y = -log10(padj))) +
     geom_point(aes(colour = padj < 0.01), size=1) +
     labs(x = "log2(fold change)", y = "-log10(p-value)", colour = "FDR < 1%",
                   title = "Acute Diarrhea vs Healthy") +
     theme_bw() +
     scale_color_manual(values = c("gray", "skyblue2"))
  
  diff_abun_ah <- sum(res1_ah$log2FoldChange > 1 & res1_ah$padj < 0.01) + sum(res1_ah$log2FoldChange < -1 & res1_ah$padj < 0.01) 
  #334 (3+331)
```

```{r vol_cdi_h, echo=FALSE, warning=FALSE, message=FALSE, fig.width=5, fig.height=4}
  res1_ch <- results(ds_multi, name=("disease_tidy_CDI_vs_healthy"), alpha = 0.05)
  res1_ch[is.na(res1_ch$padj),] <- 1

  ggplot(res1_ch, aes(x = log2FoldChange, y = -log10(padj))) +
     geom_point(aes(colour = padj < 0.01), size=1) +
     labs(x = "log2(fold change)", y = "-log10(p-value)", colour = "FDR < 1%",
                   title = "CDI vs Healthy") +
     theme_bw() +
     scale_color_manual(values = c("gray", "skyblue2"))
  
  diff_abun_ch <- sum(res1_ch$log2FoldChange > 1 & res1_ch$padj < 0.01) + sum(res1_ch$log2FoldChange < -1 & res1_ch$padj < 0.01)
  #162(1+161)
```


```{r vol_ibd_h, echo=FALSE, warning=FALSE, message=FALSE, fig.width=5, fig.height=4}
  res1_ih <- results(ds_multi, name=("disease_tidy_IBD_vs_healthy"), alpha = 0.05)
  res1_ih[is.na(res1_ih$padj),] <- 1

  ggplot(res1_ih, aes(x = log2FoldChange, y = -log10(padj))) +
     geom_point(aes(colour = padj < 0.01), size=1) +
     labs(x = "log2(fold change)", y = "-log10(p-value)", colour = "FDR < 1%",
                   title = "IBD vs Healthy") +
     theme_bw() +
     scale_color_manual(values = c("gray", "skyblue2"))
  diff_abun_ih <- sum(res1_ih$log2FoldChange > 1 & res1_ih$padj < 0.01) + sum(res1_ih$log2FoldChange < -1 & res1_ih$padj < 0.01)
  #225 (120+105)
```

`Which comparison has the highest number of differentially abundant species?`

IBD and Acute Diarrhea has the highest number of differentially abundant species with 357 species in total. 

## DESeq2 identify differential abundance species with two factors model

Here we further tested the differently abundant species between __healthy__ and __non-healthy__ patients in each __country__. Our model will be `~ age_category + disease_tidy`.


```{r deseq_two_factors_multi, echo=TRUE, warning=FALSE, message=FALSE, eval = FALSE}
  filter_lst <- c("acute_diarrhoea", "CDI", "healthy", "IBD")
  
  meta <- meta[meta$disease_tidy %in% filter_lst, ]
  meta$disease_tidy <- factor(meta$disease_tidy, levels = c("healthy", "acute_diarrhoea", "CDI", "IBD"))
  meta$age_category <- factor(meta$age_category, levels = c("adult", "child", "schoolage", "senior"))
  meta$country <- factor(meta$country, levels = c("BGD", "USA", "DNK", "ESP","NLD", "CAN", "Unknown"))
  meta$country[is.na(meta$country)] <- "Unknown" #Replacing "NA" 
  ra <- ra[, colnames(ra) %in% rownames(meta)]
  ra <- ra[rowSums(ra) > 5, ]
  # 1015 3168
  
  ## construct DESeq DataSet from matrix
  
  dds_two_multi <- DESeqDataSetFromMatrix(countData=ra, 
                              colData=meta, 
                              design=~ age_category + disease_tidy + country )


  ds_two_multi = DESeq(dds_two_multi, fitType="local", sfType = "poscounts", parallel = T)


  saveRDS(ds_two_multi, "20241006_DESeq_disease_two_factors_multi_levels.rds")
```

```{r vol_tm, echo=FALSE, warning=FALSE, message=FALSE, fig.width=5, fig.height=4}
  ds_two_multi <- readRDS("20241006_DESeq_disease_two_factors_multi_levels.rds")
  resultsNames(ds_two_multi)
```
```{r vol_ac_h, echo=FALSE, warning=FALSE, message=FALSE, fig.width=5, fig.height=4}
  res1_ac <- results(ds_two_multi, name=("age_category_child_vs_adult"), alpha = 0.05)
  res1_ac[is.na(res1_ac$padj),] <- 1

  ggplot(res1_ac, aes(x = log2FoldChange, y = -log10(padj))) +
     geom_point(aes(colour = padj < 0.01), size=1) +
     labs(x = "log2(fold change)", y = "-log10(p-value)", colour = "FDR < 1%",
                   title = "Child vs Adult") +
     theme_bw() +
     scale_color_manual(values = c("gray", "skyblue2"))
  diff_abun_ac <- sum(res1_ac$log2FoldChange > 1 & res1_ac$padj < 0.01) + sum(res1_ac$log2FoldChange < -1 & res1_ac$padj < 0.01)
  #318 (38+280)
```

```{r vol_sa_h, echo=FALSE, warning=FALSE, message=FALSE, fig.width=5, fig.height=4}
  res1_sa <- results(ds_two_multi, name=("age_category_senior_vs_adult"), alpha = 0.05)
  res1_sa[is.na(res1_sa$padj),] <- 1

  ggplot(res1_sa, aes(x = log2FoldChange, y = -log10(padj))) +
     geom_point(aes(colour = padj < 0.01), size=1) +
     labs(x = "log2(fold change)", y = "-log10(p-value)", colour = "FDR < 1%",
                   title = "Senior vs Adult") +
     theme_bw() +
     scale_color_manual(values = c("gray", "skyblue2"))
  diff_abun_sa <- sum(res1_sa$log2FoldChange > 1 & res1_sa$padj < 0.01) + sum(res1_sa$log2FoldChange < -1 & res1_sa$padj < 0.01)
  #312 (86+225)
```
