---
title: "20240929_funcitonal_profile_comparison"
output: html_document
date: "`r Sys.Date()`"
---

```{r setup, include=FALSE}
  knitr::opts_chunk$set(echo = TRUE)
  library(curatedMetagenomicData)
  library(pheatmap)
  library(Hmisc)
  library(stringr)
  library(cowplot)
  library(ggplot2)
  theme_set(theme_bw())
```

## Introduction

Firstly we import the diarrhea v4 data we exported last time. Due to the large size of functional profiling file, we need to further tidy up the data for analysis.

```{r data_import_and_filter, echo=FALSE, warning=FALSE, message=FALSE, eval = FALSE}
  dia_v4_pathway <- readRDS("20240922_diarrhea_func_profile.rds")
  
  # remove unmapped and ungrouped reads
  pathway_mat <- assay(dia_v4_pathway)
  pathway_mat <- pathway_mat[!stringr::str_detect(rownames(pathway_mat), "UNMAPPED|UNINTEGRATED"), ] 
  # 48701  3345
  
  # remove the pathways assigned to taxa and unclassified
  pathway_mat <- pathway_mat[!stringr::str_detect(rownames(pathway_mat), '__|unclassified'), ]   
  # 592 3345
  saveRDS(pathway_mat, "20240921_diarrhea_v4_func_filtered_profile.rds")
  metadata <- colData(dia_v4_pathway)
  
  metadata$Sample_ID <- rownames(metadata)
  saveRDS(metadata, "20240921_diarrhea_v4_func_filtered_profile_meta.rds")
```

### The most abundant and prevalent functional pathways in healthy and IBD samples

```{r pre_ra_separate, echo = FALSE, warning=FALSE, message=FALSE, fig.width=7, fig.height=5}
  metadata <- readRDS("20240921_diarrhea_v4_func_filtered_profile_meta.rds")
  ra_mat <- readRDS("20240921_diarrhea_v4_func_filtered_profile.rds")
 
  hc <- metadata[metadata$disease == "healthy", ]
  nhc <- metadata[metadata$disease == "IBD", ]
  
  ### Get the relative abundance (ra) and prevalence / occupancy (occu) of healthy samples
  ra_mat_hc <- ra_mat[, colnames(ra_mat) %in% hc$subject_id]
  
  ra_occu_hc <- data.frame(Pathway = rownames(ra_mat_hc),
                      RA_mean = rowMeans(ra_mat_hc),
                      RA_median = rowMedians(ra_mat_hc),
                      Prevalence = rowSums(ra_mat_hc>0)/ncol(ra_mat_hc),
                      Healthy = "HC")
  
  ### Get the relative abundance (ra) and prevalence / occupancy (occu) of non-healthy samples
  ra_mat_nhc <- ra_mat[, colnames(ra_mat) %in% nhc$subject_id]
  
  ra_occu_nhc <- data.frame(Pathway = rownames(ra_mat_nhc),
                      RA_mean = rowMeans(ra_mat_nhc),
                      RA_median = rowMedians(ra_mat_nhc),
                      Prevalence = rowSums(ra_mat_nhc>0)/ncol(ra_mat_nhc),
                      Healthy = "IBD")
  
  ra_occu <- rbind(ra_occu_hc, ra_occu_nhc)
  
  ggplot(ra_occu, aes(RA_mean, Prevalence)) + 
    geom_point(shape = 1, color = "gray") + 
    facet_wrap(~Healthy, scales = "free") +
    labs(x = "Average relative abundance") + 
    theme(legend.position = "bottom") +
    guides(col = guide_legend(nrow = 3))
```

Unlike species, we observe that many function pathways have high prevalence. Here we plot the distribution of Prevalence and RA between healthy and non-healthy groups.

```{r dis_prevalence, echo = FALSE, warning=FALSE, message=FALSE, fig.width=6, fig.height=2}
  ggplot(ra_occu, aes(Prevalence)) + 
    geom_histogram(aes(fill = Healthy), position = "dodge") +
    scale_fill_manual(values = c("salmon", "darkolivegreen3")) +
    ylab("Count")
```

From this distribution, we observed that in healthy sample, more unique pathways were found and in IBD samples, more common pathways were identified.

```{r dis_ra, echo = FALSE, warning=FALSE, message=FALSE, fig.width=6, fig.height=4}
  p1 <- ggplot(ra_occu, aes(RA_mean)) + 
    geom_histogram(aes(fill = Healthy), position = "dodge") +
    scale_fill_manual(values = c("salmon", "darkolivegreen3")) +
    ylab("Count")
  p2 <- ggplot(ra_occu, aes(RA_median)) + 
    geom_histogram(aes(fill = Healthy), position = "dodge") +
    scale_fill_manual(values = c("salmon", "darkolivegreen3")) +
    ylab("Count")
  plot_grid(p1, p2, nrow = 2)
```

```
Plot the prevalence and mean/median relative abundance (RA) of taxonomic disbritubion of HC and IBD samples, what do you observe?
```

### Correlation between functional pathways 

Get the correlation coefficient adjacency matrix and plot heatmap.

```{r filter_rare, echo=FALSE, warning=FALSE, message=FALSE}
  n_sample <- ncol(ra_mat)
  n_sample_filter <- round(n_sample/20)
  
  ra_mat_filter1 <- ra_mat[rowSums(ra_mat > 0) > n_sample_filter, ]
  ra_mat_filter2 <- ra_mat[rowSums(ra_mat) / n_sample > 0.00001, ]
  
  pass_filter <- intersect(rownames(ra_mat_filter1), rownames(ra_mat_filter2))
  
  ra_mat_filter <- ra_mat[pass_filter, ]
  
  #dim(ra_mat_filter)
  
  ## re-normalize the relative abundance by column
  ra_mat_filter <-  as.matrix(apply(ra_mat_filter, 2, function(x) x/sum(x)))
  
```

```{r corr, echo=FALSE, warning=FALSE, message=FALSE, fig.width=26, fig.height=24}
  ra_corr <- rcorr(t(ra_mat_filter), type = "spearman")
  
  ## correlation matrix
  mat_cor <- ra_corr$r
  ## significance matrix
  sig_cor <- ra_corr$P
  
  ## filter out the strong (abs > 0.6) and significant (P < 0.01) correlation
  mat_cor[abs(mat_cor) < 0.6] <- 0
  mat_cor[sig_cor > 0.01] <- 0
  diag(mat_cor) <- 0 
  
  mat_cor_filter <- mat_cor[rowSums(mat_cor) > 0, colSums(mat_cor) > 0]
  
  #heatmap(mat_cor_filter, symm = TRUE, margins = c(6,6))
  
  pheatmap(mat_cor_filter, na_col="white",
          border_color ="NA",  treeheight_row = 0, treeheight_col = 0)
  
  
```

#### Assignment
```{r corr cmp h, echo=FALSE, warning=FALSE, message=FALSE, fig.width=26, fig.height=24}

  hc <- metadata[metadata$disease == "healthy", ]
  nhc <- metadata[metadata$disease == "IBD", ]
  
  ra_mat_hc <- as.data.frame(ra_mat_filter[, colnames(ra_mat_filter) %in% hc$subject_id])
  ra_corr_hc <- rcorr(t(ra_mat_hc), type = "spearman")
  
  ## correlation matrix
  mat_cor_hc <- ra_corr_hc$r
  ## significance matrix
  sig_cor_hc <- ra_corr_hc$P
  
  ## filter out the strong (abs > 0.6) and significant (P < 0.01) correlation
  mat_cor_hc[abs(mat_cor_hc) < 0.6] <- 0
  mat_cor_hc[sig_cor_hc > 0.01] <- 0
  diag(mat_cor_hc) <- 0 
  
  mat_cor_filter_hc <- mat_cor_hc[rowSums(mat_cor_hc) > 0, colSums(mat_cor_hc) > 0]
  
  p1 <- pheatmap(mat_cor_filter_hc, na_col="white",
          border_color ="NA",  treeheight_row = 0, treeheight_col = 0, main = "Healthy Functional Pathway")
  
```

```{r corr cmp ibd, warning=FALSE, echo=FALSE, message=FALSE, fig.width=26, fig.height=24}
  
  
  ra_mat_nhc <- as.data.frame(ra_mat_filter[, colnames(ra_mat_filter) %in% nhc$subject_id])
  ra_corr_nhc <- rcorr(t(ra_mat_nhc), type = "spearman")
  
  ## correlation matrix
  mat_cor_nhc <- ra_corr_nhc$r
  ## significance matrix
  sig_cor_nhc <- ra_corr_nhc$P
  
  ## filter out the strong (abs > 0.6) and significant (P < 0.01) correlation
  mat_cor_nhc[abs(mat_cor_nhc) < 0.6] <- 0
  mat_cor_nhc[sig_cor_nhc > 0.01] <- 0
  diag(mat_cor_nhc) <- 0 
  
  mat_cor_filter_nhc <- mat_cor_nhc[rowSums(mat_cor_nhc) > 0, colSums(mat_cor_nhc) > 0]
  
  #heatmap(mat_cor_filter, symm = TRUE, margins = c(6,6))
  
  p2 <- pheatmap(mat_cor_filter_nhc, na_col="white",
          border_color ="NA",  treeheight_row = 0, treeheight_col = 0, main = "IBD Functional Pathway")
  
```
Compare the pathway interaction heatmap between IBD and healthy samples.
